src/visualizers/panels/ExecutionIndex/ExecutionIndexControl.js

```diff
@@ -38,45 +41,88 @@ define([
-                .reduce((l1, l2) => l1.concat(l2), []);
-
-            this._updateLines(otherLines, false);
-            this._updateLines(otherLines, true);
+        let displayedGraphExecIds = Object.keys(this.displayedExecutions)
+            .filter((id) => !!this.displayedExecutions[id]);
+        if (this._currentPlotsDataId) {
+            this._widget.removeNode(this._currentPlotsDataId);
+        }
+        if ( this.displayedExecCount() > 0) {
+            this._consolidateGraphData(displayedGraphExecIds);
         }
-
-        this._updateLines(lines, bool);
     };
 
-    ExecutionIndexControl.prototype._updateLines = function (lines, added) {
-        var action = added ? 'addNode' : 'removeNode';
+    ExecutionIndexControl.prototype._consolidateGraphData = function (graphExecIDs) {
+        let graphIds = graphExecIDs.map(execId => this._graphsForExecution[execId]);
+        let graphDescs = graphIds.map(id => this._getObjectDescriptor(id)).filter(desc => !!desc);
+        if(graphDescs.length > 0){
+            let consolidatedDesc = this._combineGraphDesc(graphDescs);
+            this._currentPlotsDataId = consolidatedDesc.id;
+            let plotlyJSON = this._plotlyDescExtractor.descToPlotlyJSON(consolidatedDesc);
+            plotlyJSON.type = 'graph';
+            this._widget.addNode(plotlyJSON);
+        }
+    };
 
-        // If removing, just get the ids
-        lines = !added ? lines : lines.map(line => this._getObjectDescriptor(line))
-            .filter(line => !!line);
+    ExecutionIndexControl.prototype._combineGraphDesc = function (graphDescs) {
+        const isMultiGraph = this.displayedExecCount() > 1;
+        if (!isMultiGraph) {
+            return graphDescs[0];
+        } else {
+            let consolidatedDesc = null;
+
+            graphDescs.forEach((desc) => {
+                if (!consolidatedDesc) {
+                    consolidatedDesc = JSON.parse(JSON.stringify(desc));
+                    consolidatedDesc.title = consolidatedDesc.title || `Graph`;
+                } else {
+                    consolidatedDesc.id += desc.id;
+                    consolidatedDesc.execId += ` vs ${desc.execId}`;
+                    consolidatedDesc.graphId += ` vs ${appendStringWithAbbr(desc.execId, desc.abbr)}`;
+                    consolidatedDesc.title +=
+                        ` vs ${appendStringWithAbbr(desc.title || 'Graph', desc.abbr)}`;
+                    this._combineSubGraphsDesc(consolidatedDesc, desc.subGraphs, desc.abbr);
+                }
+            });
+            return consolidatedDesc;
+        }
+    };
 
-        // update the given lines
-        for (var i = lines.length; i--;) {
-            this._widget[action](lines[i]);
+    ExecutionIndexControl.prototype._combineSubGraphsDesc = function(consolidatedDesc, subGraphs, abbr){
+        let currentSubGraph;
+        for(let i = 0; i < consolidatedDesc.subGraphs.length; i++){
+            if(!subGraphs[i]) break;
```

Comment:
Can this happen? I have seen a few places where the code is pretty pessimistic about if the graph/subgraph descriptor exists. It might be nice to just ensure that they all do exist (if possible).

Reply:
This method tries to combine all the `SubGraphs`. Based on our previous conversations, There can be two possible scenarios. `existingGraph` has more `subGraph` than the `addedGraph`. Added graph has more `subGraphs` than `existingGraph`. So, this for loop should break when `addedGraph` has less `subGraphs` than the `exisitinGraph`